keanu-project/src/main/java/io/improbable/keanu/vertices/generic/nonprobabilistic/If.java

```diff
@@ -63,8 +61,8 @@ public IfThenElseBuilder(Vertex<? extends BooleanTensor> predicate,
         }
 
         public IfVertex<T> orElse(Vertex<? extends Tensor<T>> els) {
-            assertShapesMatchOrAreScalar(thn.getShape(), els.getShape(), predicate.getShape());
-            return new IfVertex<>(els.getShape(), predicate, thn, els);
+            long[] shape = TensorShapeValidation.checkHasOneNonSingularShapeOrAllSingular(thn.getShape(), els.getShape(), predicate.getShape());
```

Comment:
I don't think we can use `checkHasOneNonSingularShapeOrAllSingular` as it will pass if the `then` condition is non singular and the `else` and `predicate` are singular - which is wrong

Reply:
Add a check to make sure thn and els are always the same shape.